File: systems/plants/constraint/IKoptions.m
classdef IKoptions
% @param Q                  --the cost matrix for posture deviation
% @param use_mex,           -- A boolean flat,set to true if run with mex. default is
%                              false
% @param debug_mode         -- IK would query the name of each constraint
%
% The following fields is used in inverseKinPointwise
% @param sequentialSeedFlag -- A boolean variable, set to false if the seed of
%                             inverseKinPointwise is q_seed. If set to true,
%                             inverseKinPointwise would use the solution in the
%                             previous time point as seed (if it is feasible).
%
% All subsequent fields will be used in inverseKinTraj only
% @param Qa                 -- The cost matrix for acceleration
% @param Qv                 -- The cost matrix for velocity
% @param additional_tSamples  -- Additional time samples (apart from knot points) in inverseKinTraj to check
%                                constraints
% @param fixInitialState    -- A boolean flag, set to true if the posture q and
%                              velocity qd at time t(1) is fixed. q(1) = q_seed_traj.eval(t(1))
%                              qdot(1) = (ikpotions.qd0_lb+ikoptions.qd0_ub)/2
% @param q0_lb              -- The initial posture upperbound. Would be used when
%                              fixInitialState = false;
% @param q0_ub              -- The initial posture lowerbound, Would be used when
%                              fixInitialState = false;
% @param qd0_ub             -- The upperbound of the initial velocity
% @param qd0_lb             -- The lowerbound of the initial velocity
% @param qdf_ub             -- The upper bound of the final velocity
% @param qdf_lb             -- The lower bound of the final velocity
  properties(SetAccess=protected)
    robot
    nq
    Q
    Qa
    Qv
    use_mex
    debug_mode
    sequentialSeedFlag
    SNOPT_MajorFeasibilityTolerance
    SNOPT_MajorIterationsLimit
    SNOPT_IterationsLimit
    SNOPT_SuperbasicsLimit
    SNOPT_MajorOptimalityTolerance
    additional_tSamples
    fixInitialState
    q0_lb
    q0_ub
    qd0_lb
    qd0_ub
    qdf_ub
    qdf_lb
  end
  methods
    function obj = IKoptions(robot)
      obj.robot = robot;
      obj.nq = obj.robot.getNumDOF();
      obj.Q = eye(obj.nq);
      obj.Qa = 0.1*eye(obj.nq);
      obj.Qv = zeros(obj.nq);
%       obj.approximateFlag = false;
      obj.use_mex = true;
      obj.debug_mode = false;
      obj.sequentialSeedFlag = false;
      obj.SNOPT_MajorFeasibilityTolerance = 1e-6;
      obj.SNOPT_MajorIterationsLimit = 200;
      obj.SNOPT_IterationsLimit = 10000;
      obj.SNOPT_SuperbasicsLimit = 2000;
      obj.SNOPT_MajorOptimalityTolerance = 1e-4;
      obj.additional_tSamples = [];
      obj.fixInitialState = true;
      [obj.q0_lb,obj.q0_ub] = obj.robot.getJointLimits();
      obj.qd0_ub = zeros(obj.nq,1);
      obj.qd0_lb = zeros(obj.nq,1);
      obj.qdf_ub = zeros(obj.nq,1);
      obj.qdf_lb = zeros(obj.nq,1);
    end
    
    function obj = setQ(obj,Q)
      typecheck(Q,'double');
      sizecheck(Q,[obj.nq,obj.nq]);
      Q = (Q+Q')/2;
      if(any(abs(eig(Q))<-1e-10))
        error('IKoptions:Q must be positive semidefinite');
      end
      obj.Q = Q;
    end
    
    function obj = setQa(obj,Qa)
      typecheck(Qa,'double');
      sizecheck(Qa,[obj.nq,obj.nq]);
      Qa = (Qa+Qa')/2;
      if(any(abs(eig(Qa))<-1e-10))
        error('IKoptions:Qa must be positive semidefinite');
      end
      obj.Qa = Qa;
    end
    
    function obj = setQv(obj,Qv)
      typecheck(Qv,'double');
      sizecheck(Qv,[obj.nq,obj.nq]);
      Qv = (Qv+Qv')/2;
      if(any(abs(eig(Qv))<-1e-10))
        error('IKoptions:Qv must be positive semidefinite');
      end
      obj.Qv = Qv;
    end
    
    function obj = setMex(obj,flag)
      sizecheck(flag,[1,1]);
      if(isa(flag,'logical'))
        obj.use_mex = flag;
      elseif(isa(flag,'double'))
        obj.use_mex = logical(flag);
      else
        error('IKoptions: Unsupported flag type');
      end
    end
    
    function obj = setDebug(obj,flag)
      sizecheck(flag,[1,1]);
      if(isa(flag,'logical'))
        obj.debug_mode = flag;
      elseif(isa(flag,'double'))
        obj.debug_mode = logical(flag);
      else
        error('IKoptions: Unsupported flag type');
      end
    end
    
    function obj = setSequentialSeedFlag(obj,flag)
      sizecheck(flag,[1,1]);
      if(isa(flag,'logical'))
        obj.sequentialSeedFlag = flag;
      elseif(isa(flag,'double'))
        obj.sequentialSeedFlag = logical(flag);
      else
        error('Drake:IKoptions: flag must be a logical scalar');
      end
    end
    
    function obj = setMajorOptimalityTolerance(obj,tol)
      sizecheck(tol,[1,1]);
      typecheck(tol,'double');
      if(tol<=0)
        error('Drake:IKoptions: tol must be positive');
      end
      obj.SNOPT_MajorOptimalityTolerance = tol;
    end
    
    function obj = setMajorFeasibilityTolerance(obj,tol)
      sizecheck(tol,[1,1]);
      typecheck(tol,'double');
      if(tol<=0)
        error('Drake:IKoptions: tol must be positive');
      end
      obj.SNOPT_MajorFeasibilityTolerance = tol;
    end
    
    function obj = setSuperbasicsLimit(obj,limit)
      sizecheck(limit,[1,1]);
      typecheck(limit,'double');
      if(limit<1)
        error('Drake:IKoptions: limit must be positive integers');
      end
      obj.SNOPT_SuperbasicsLimit = floor(limit);
    end
    
    function obj = setMajorIterationsLimit(obj,limit)
      sizecheck(limit,[1,1]);
      typecheck(limit,'double');
      if(limit<1)
        error('Drake:IKoptions: limit must be positive integers');
      end
      obj.SNOPT_MajorIterationsLimit = floor(limit);
    end
    
    function obj = setIterationsLimit(obj,limit)
      sizecheck(limit,[1,1]);
      typecheck(limit,'double');
      if(limit<1)
        error('Drake:IKoptions: limit must be positive integers');
      end
      obj.SNOPT_IterationsLimit = floor(limit);
    end
    
    function obj = setFixInitialState(obj,flag)
      sizecheck(flag,[1,1]);
      if(isa(flag,'logical'))
        obj.fixInitialState = flag;
      elseif(isa(flag,'double'))
        obj.fixInitialState = logical(flag);
      else
        error('IKoptions: Unsupported flag type');
      end
    end
    function obj = setq0(obj,lb,ub)
      typecheck(lb,'double');
      typecheck(ub,'double');
      sizecheck(lb,[obj.nq,1]);
      sizecheck(ub,[obj.nq,1]);
      if(any(lb>ub))
        error('IKoptions:setq0: lb must be no larger than ub');
      end
      [lb_tmp,ub_tmp] = obj.robot.getJointLimits();
      obj.q0_lb = max([lb lb_tmp],[],2);
      obj.q0_ub = min([ub ub_tmp],[],2);
    end
    
    function obj = setqd0(obj,lb,ub)
      typecheck(lb,'double');
      typecheck(ub,'double');
      sizecheck(lb,[obj.nq,1]);
      sizecheck(ub,[obj.nq,1]);
      if(any(lb>ub))
        error('IKoptions:setqd0: lb must be no larger than ub');
      end
      obj.qd0_lb = lb;
      obj.qd0_ub = ub;
    end
    
    
    function obj = setqdf(obj,lb,ub)
      typecheck(lb,'double');
      typecheck(ub,'double');
      sizecheck(lb,[obj.nq,1]);
      sizecheck(ub,[obj.nq,1]);
      if(any(lb>ub))
        error('IKoptions:setqdf: lb must be no larger than ub');
      end
      obj.qdf_lb = lb;
      obj.qdf_ub = ub;
    end

    function obj = setAdditionaltSamples(obj,t_samples)
      typecheck(t_samples,'double');
      if(~isempty(t_samples))
        sizecheck(t_samples,[1,nan]);
        obj.additional_tSamples = unique(t_samples);
      else
        obj.additional_tSamples = [];
      end
    end

    function obj = updateRobot(obj,robot)
      obj.robot = robot;
      nq_cache = obj.nq;
      obj.nq = obj.robot.getNumDOF();
      if(obj.nq ~= nq_cache)
        obj.Q = eye(obj.nq);
        obj.Qa = 0.1*eye(obj.nq);
        obj.Qv = zeros(obj.nq);
        [obj.q0_lb,obj.q0_ub] = obj.robot.getJointLimits();
        obj.qd0_ub = zeros(obj.nq,1);
        obj.qd0_lb = zeros(obj.nq,1);
        obj.qdf_ub = zeros(obj.nq,1);
        obj.qdf_lb = zeros(obj.nq,1);
      end
    end
  end
end

File: drake/systems/plants/IKoptions.cpp
#include <set>

#include "drake/systems/plants/RigidBodyTree.h"
#include "drake/systems/plants/IKoptions.h"

using Eigen::MatrixXd;
using Eigen::RowVectorXd;
using Eigen::SelfAdjointEigenSolver;
using Eigen::VectorXd;
using std::cerr;
using std::endl;
using std::set;

IKoptions::IKoptions(RigidBodyTree *robot) {
  // It is important to make sure these default values are consistent with the
  // MATLAB IKoptions
  setDefaultParams(robot);
}

IKoptions::IKoptions(const IKoptions &rhs) {
  robot_ = rhs.robot_;
  nq_ = rhs.nq_;
  Q_ = rhs.Q_;
  Qa_ = rhs.Qa_;
  Qv_ = rhs.Qv_;
  debug_mode_ = rhs.debug_mode_;
  sequentialSeedFlag_ = rhs.sequentialSeedFlag_;
  SNOPT_MajorFeasibilityTolerance_ = rhs.SNOPT_MajorFeasibilityTolerance_;
  SNOPT_MajorIterationsLimit_ = rhs.SNOPT_MajorIterationsLimit_;
  SNOPT_IterationsLimit_ = rhs.SNOPT_IterationsLimit_;
  SNOPT_SuperbasicsLimit_ = rhs.SNOPT_SuperbasicsLimit_;
  SNOPT_MajorOptimalityTolerance_ = rhs.SNOPT_MajorOptimalityTolerance_;
  additional_tSamples_ = rhs.additional_tSamples_;
  fixInitialState_ = rhs.fixInitialState_;
  q0_lb_ = rhs.q0_lb_;
  q0_ub_ = rhs.q0_ub_;
  qd0_lb_ = rhs.qd0_lb_;
  qd0_ub_ = rhs.qd0_ub_;
  qdf_lb_ = rhs.qdf_lb_;
  qdf_ub_ = rhs.qdf_ub_;
}
IKoptions::~IKoptions() {}

void IKoptions::setDefaultParams(RigidBodyTree *robot) {
  robot_ = robot;
  nq_ = robot->get_num_positions();
  Q_ = MatrixXd::Identity(nq_, nq_);
  Qa_ = 0.1 * MatrixXd::Identity(nq_, nq_);
  Qv_ = MatrixXd::Zero(nq_, nq_);
  debug_mode_ = true;
  sequentialSeedFlag_ = false;
  SNOPT_MajorFeasibilityTolerance_ = 1E-6;
  SNOPT_MajorIterationsLimit_ = 200;
  SNOPT_IterationsLimit_ = 10000;
  SNOPT_SuperbasicsLimit_ = 2000;
  SNOPT_MajorOptimalityTolerance_ = 1E-4;
  additional_tSamples_.resize(0);
  fixInitialState_ = true;
  q0_lb_ = robot->joint_limit_min;
  q0_ub_ = robot->joint_limit_max;
  qd0_ub_ = VectorXd::Zero(nq_);
  qd0_lb_ = VectorXd::Zero(nq_);
  qdf_ub_ = VectorXd::Zero(nq_);
  qdf_lb_ = VectorXd::Zero(nq_);
}

RigidBodyTree *IKoptions::getRobotPtr() const { return robot_; }

void IKoptions::setQ(const MatrixXd &Q) {
  if (Q.rows() != nq_ || Q.cols() != nq_) {
    cerr << "Q should be nq x nq matrix" << endl;
  }
  Q_ = (Q + Q.transpose()) / 2;
  SelfAdjointEigenSolver<MatrixXd> eigensolver(Q_);
  VectorXd ev = eigensolver.eigenvalues();
  for (int i = 0; i < nq_; i++) {
    if (ev(i) < 0) {
      cerr << "Q is not positive semi-definite" << endl;
    }
  }
}

void IKoptions::setQa(const MatrixXd &Qa) {
  if (Qa.rows() != nq_ || Qa.cols() != nq_) {
    cerr << "Qa should be nq x nq matrix" << endl;
  }
  Qa_ = (Qa + Qa.transpose()) / 2;
  SelfAdjointEigenSolver<MatrixXd> eigensolver(Qa_);
  VectorXd ev = eigensolver.eigenvalues();
  for (int i = 0; i < nq_; i++) {
    if (ev(i) < 0) {
      cerr << "Qa is not positive semi-definite" << endl;
    }
  }
}

void IKoptions::setQv(const MatrixXd &Qv) {
  if (Qv.rows() != nq_ || Qv.cols() != nq_) {
    cerr << "Qv should be nq x nq matrix" << endl;
  }
  Qv_ = (Qv + Qv.transpose()) / 2;
  SelfAdjointEigenSolver<MatrixXd> eigensolver(Qv_);
  VectorXd ev = eigensolver.eigenvalues();
  for (int i = 0; i < nq_; i++) {
    if (ev(i) < 0) {
      cerr << "Qv is not positive semi-definite" << endl;
    }
  }
}
void IKoptions::getQ(MatrixXd &Q) const { Q = Q_; }

void IKoptions::getQa(MatrixXd &Qa) const { Qa = Qa_; }

void IKoptions::getQv(MatrixXd &Qv) const { Qv = Qv_; }

void IKoptions::setDebug(bool flag) { debug_mode_ = flag; }

bool IKoptions::getDebug() const { return debug_mode_; }

void IKoptions::setSequentialSeedFlag(bool flag) {
  sequentialSeedFlag_ = flag;
}

bool IKoptions::getSequentialSeedFlag() const {
  return sequentialSeedFlag_;
}

void IKoptions::setMajorOptimalityTolerance(double tol) {
  if (tol <= 0) {
    cerr << "Major Optimality Tolerance must be positive" << endl;
  }
  SNOPT_MajorOptimalityTolerance_ = tol;
}

double IKoptions::getMajorOptimalityTolerance() const {
  return SNOPT_MajorOptimalityTolerance_;
}

void IKoptions::setMajorFeasibilityTolerance(double tol) {
  if (tol <= 0) {
    cerr << "Major Feasibility Tolerance must be positive" << endl;
  }
  SNOPT_MajorFeasibilityTolerance_ = tol;
}

double IKoptions::getMajorFeasibilityTolerance() const {
  return SNOPT_MajorFeasibilityTolerance_;
}

void IKoptions::setSuperbasicsLimit(int limit) {
  if (limit <= 0) {
    cerr << "Superbasics limit must be positive" << endl;
  }
  SNOPT_SuperbasicsLimit_ = limit;
}

int IKoptions::getSuperbasicsLimit() const {
  return SNOPT_SuperbasicsLimit_;
}

void IKoptions::setMajorIterationsLimit(int limit) {
  if (limit <= 0) {
    cerr << "Major iterations limit must be positive" << endl;
  }
  SNOPT_MajorIterationsLimit_ = limit;
}

int IKoptions::getMajorIterationsLimit() const {
  return SNOPT_MajorIterationsLimit_;
}

void IKoptions::setIterationsLimit(int limit) {
  if (limit <= 0) {
    cerr << "Iterations limit must be positive" << endl;
  }
  SNOPT_IterationsLimit_ = limit;
}

int IKoptions::getIterationsLimit() const {
  return SNOPT_IterationsLimit_;
}

void IKoptions::setFixInitialState(bool flag) { fixInitialState_ = flag; }

bool IKoptions::getFixInitialState() const { return fixInitialState_; }

void IKoptions::setq0(const VectorXd &lb, const VectorXd &ub) {
  if (lb.rows() != nq_ || ub.rows() != nq_) {
    cerr << "q0_lb and q0_ub must be nq x 1 column vector" << endl;
  }
  q0_lb_ = lb;
  q0_ub_ = ub;
  for (int i = 0; i < nq_; i++) {
    if (q0_lb_(i) > q0_ub_(i)) {
      cerr << "q0_lb must be no larger than q0_ub" << endl;
    }
    q0_lb_(i) = q0_lb_(i) > robot_->joint_limit_min[i]
                ? q0_lb_(i)
                : robot_->joint_limit_min[i];
    q0_ub_(i) = q0_ub_(i) < robot_->joint_limit_max[i]
                ? q0_ub_(i)
                : robot_->joint_limit_max[i];
  }
}

void IKoptions::getq0(VectorXd &lb, VectorXd &ub) const {
  lb = q0_lb_;
  ub = q0_ub_;
}

void IKoptions::setqd0(const VectorXd &lb, const VectorXd &ub) {
  if (lb.rows() != nq_ || ub.rows() != nq_) {
    cerr << "qd0_lb and qd0_ub must be nq x 1 column vector" << endl;
  }
  for (int i = 0; i < nq_; i++) {
    if (lb(i) > ub(i)) {
      cerr << "qd0_lb must be no larger than qd0_ub" << endl;
    }
  }
  qd0_lb_ = lb;
  qd0_ub_ = ub;
}

void IKoptions::getqd0(VectorXd &lb, VectorXd &ub) const {
  lb = qd0_lb_;
  ub = qd0_ub_;
}

void IKoptions::setqdf(const VectorXd &lb, const VectorXd &ub) {
  if (lb.rows() != nq_ || ub.rows() != nq_) {
    cerr << "qdf_lb and qdf_ub must be nq x 1 column vector" << endl;
  }
  for (int i = 0; i < nq_; i++) {
    if (lb(i) > ub(i)) {
      cerr << "qdf_lb must be no larger than qdf_ub" << endl;
    }
  }
  qdf_lb_ = lb;
  qdf_ub_ = ub;
}

void IKoptions::getqdf(VectorXd &lb, VectorXd &ub) const {
  lb = qdf_lb_;
  ub = qdf_ub_;
}

void IKoptions::setAdditionaltSamples(const RowVectorXd &t_samples) {
  if (t_samples.size() > 0) {
    set<double> unique_sort_t(t_samples.data(),
                              t_samples.data() + t_samples.size());
    additional_tSamples_.resize(unique_sort_t.size());
    int t_idx = 0;
    for (auto it = unique_sort_t.begin(); it != unique_sort_t.end(); it++) {
      additional_tSamples_(t_idx) = *it;
      t_idx++;
    }
  } else {
    additional_tSamples_.resize(0);
  }
}

void IKoptions::getAdditionaltSamples(RowVectorXd &t_samples) const {
  t_samples = additional_tSamples_;
}

void IKoptions::updateRobot(RigidBodyTree *new_robot) {
  robot_ = new_robot;
  int nq_cache = nq_;
  nq_ = robot_->get_num_positions();
  if (nq_cache != nq_) {
    setDefaultParams(new_robot);
  }
}
